### PR TITLE
Cram size2 (htslib#1546)

### PR DESCRIPTION
**Pull request by jkbonfield**
This adds a cram_size sub-command, modelled on Io_lib's cram_size program.

It dumps a bunch of metrics about a CRAM file, focusing on (for most generated CRAMs at least) aggregate sizes per block "Content ID" fields, the data-series contained within them, and the compression methods used.

An example on a modern NovaSeq CRAM 3.1 file (10 million records):

```
#   Content_ID  Uncomp.size    Comp.size   Ratio Method  Data_series
BLOCK     CORE            0            0 100.00% .      
BLOCK       11    394734019     31058554   7.87% n       RN
BLOCK       12   1504781763     80451466   5.35% f       QS
BLOCK       13       330065        63255  19.16% _g.0    IN
BLOCK       14     26625602      6725907  25.26% 111     SC
BLOCK       15     15002423      2910416  19.40% 1       BF
BLOCK       16     10000000      1416056  14.16% 1       CF
BLOCK       17     10000383      4722460  47.22% 0       AP
BLOCK       19      9956162       452265   4.54% 0120    MQ
BLOCK       20       375940       170246  45.29% 00      NS
BLOCK       21       375940        80602  21.44% 01      MF
BLOCK       22       588539       279419  47.48% _g      TS
BLOCK       23      1485289      1088800  73.31% _g      NP
BLOCK       24      5139008      4362646  84.89% _bg     NF
BLOCK       25     19952610       131643   0.66% _b2     RL
BLOCK       26      9956162      2709414  27.21% 0       FN
BLOCK       27     12447702       823667   6.62% 10      FC
BLOCK       28     13174993     10283151  78.05% 0       FP
BLOCK       29       148413        45277  30.51% 002g0   DL
BLOCK       30      6694464      1523063  22.75% 1       BA
BLOCK       31     11341391      1954969  17.24% 1       BS
BLOCK       32     10000000       174095   1.74% 10      TL
BLOCK       36        79026        77762  98.40% .0g     HC
BLOCK  4281155     10000000      3935234  39.35% 0g      ASC
BLOCK  5063770         1240          503  40.57% b.      MDZ
BLOCK  5131587          185           85  45.98% 0.      NMC
BLOCK  5456218      2885009      1024636  35.52% bg      SAZ
BLOCK  5783898     21340853      4633677  21.71% b       XAZ
BLOCK  5788483     10000000      4858458  48.58% 10b     XSC

Number of containers                 400
Number of slices                     400
Number of sequences             10000000
Number of bases               1504781763
Total file size                166249097
Format overhead size              291371
```

In verbose mode it expands up those codecs:

```
#   Content_ID  Uncomp.size    Comp.size   Ratio Method      Data_series
BLOCK     CORE            0            0 100.00% raw        
BLOCK       11    394734019     31058554   7.87% tok3-rans   RN
BLOCK       12   1504781763     80451466   5.35% fqzcomp     QS
BLOCK       13       319718        61183  19.14% gzip-min    IN
BLOCK       13        10088         1916  18.99% gzip        IN
BLOCK       13           86           86 100.00% raw         IN
BLOCK       13          173           70  40.50% r4x16-o0    IN
BLOCK       14     18310603      4656720  25.43% r4x16-o1    SC
BLOCK       14      7890427      1972780  25.00% r4x16-o1P   SC
BLOCK       14       424572        96407  22.71% r4x16-o1PR  SC
BLOCK       15     15002423      2910416  19.40% r4x16-o1    BF
BLOCK       16     10000000      1416056  14.16% r4x16-o1    CF
BLOCK       17     10000383      4722460  47.22% r4x16-o0    AP
BLOCK       19      8962746       360781   4.03% r4x16-o0R   MQ
BLOCK       19       471985        72249  15.31% r4x16-o1R   MQ
BLOCK       19       421824        17470   4.14% rNx16-cat   MQ
BLOCK       19        99607         1765   1.77% r4x16-o0    MQ
BLOCK       20       374846       169759  45.29% r4x16-o0    NS
BLOCK       20         1094          487  44.52% r4x16-o0R   NS
BLOCK       21       369810        79260  21.43% r4x16-o0    MF
BLOCK       21         6130         1342  21.89% r4x16-o1    MF
BLOCK       22       568761       270483  47.56% gzip-min    TS
BLOCK       22        19778         8936  45.18% gzip        TS
BLOCK       23      1438922      1055817  73.38% gzip-min    NP
BLOCK       23        46367        32983  71.13% gzip        NP
BLOCK       24      4430820      3767854  85.04% gzip-min    NF
BLOCK       24       494575       412230  83.35% bzip2-6     NF
BLOCK       24       213613       182562  85.46% gzip        NF
BLOCK       25     17307181       118324   0.68% gzip-min    RL
BLOCK       25      2595454        13222   0.51% bzip2-6     RL
BLOCK       25        49975           97   0.19% rNx16-cat   RL
BLOCK       26      9956162      2709414  27.21% r4x16-o0    FN
BLOCK       27     12416020       821783   6.62% r4x16-o1    FC
BLOCK       27        31682         1884   5.95% r4x16-o0    FC
BLOCK       28     13174993     10283151  78.05% r4x16-o0    FP
BLOCK       29        63149        18673  29.57% r4x16-o0R   DL
BLOCK       29        38399        13626  35.49% r4x16-o0    DL
BLOCK       29        41685        11710  28.09% rNx16-cat   DL
BLOCK       29         4635         1119  24.14% gzip        DL
BLOCK       29          545          149  27.35% r4x16-o0PR  DL
BLOCK       30      6694464      1523063  22.75% r4x16-o1    BA
BLOCK       31     11341391      1954969  17.24% r4x16-o1    BS
BLOCK       32      9950000       172111   1.73% r4x16-o1    TL
BLOCK       32        50000         1984   3.97% r4x16-o0PR  TL
BLOCK       36        62183        62183 100.00% raw         HC
BLOCK       36        16405        15151  92.36% r4x16-o0    HC
BLOCK       36          438          428  97.72% gzip        HC
BLOCK  4281155      9975000      3922351  39.32% r4x16-o0    ASC
BLOCK  4281155        25000        12883  51.53% gzip        ASC
BLOCK  5063770         1200          463  38.59% bzip2-6     MDZ
BLOCK  5063770           40           40 100.00% raw         MDZ
BLOCK  5131587          180           80  44.48% r4x16-o0    NMC
BLOCK  5131587            5            5 100.00% raw         NMC
BLOCK  5456218      2147878       741947  34.54% bzip2-6     SAZ
BLOCK  5456218       737131       282689  38.35% gzip        SAZ
BLOCK  5783898     21340853      4633677  21.71% bzip2-6     XAZ
BLOCK  5788483      9075000      4366510  48.12% r4x16-o1    XSC
BLOCK  5788483       900000       480424  53.38% r4x16-o0    XSC
BLOCK  5788483        25000        11524  46.10% bzip2-6     XSC

Number of containers                 400
Number of slices                     400
Number of sequences             10000000
Number of bases               1504781763
Total file size                166249097
Format overhead size              291371
```

Finally there is an `-e` option to report the per-container encodings meta-data.  This is much more verbose, and duplicates a little bit of the Io_lib `cram_dump` output.  I'm unsure if this is really beneficial or not as there is some logic to the way that cram_dump does this by decoding as much of it as possible itself without using library calls.  This ensures the data is shown as-is in the order it is, and not after it's been sanitized and reordered.  Anyway it's e.g.:

```
Container encodings
	RN	BYTE_ARRAY_STOP(stop=0,id=11)
	QS	EXTERNAL(id=12)
	IN	BYTE_ARRAY_STOP(stop=0,id=13)
	SC	BYTE_ARRAY_STOP(stop=0,id=14)
	BF	EXTERNAL(id=15)
	CF	EXTERNAL(id=16)
	AP	EXTERNAL(id=17)
	RG	HUFFMAN(codes={-1},lengths={0})
	MQ	EXTERNAL(id=19)
	NS	EXTERNAL(id=20)
	MF	EXTERNAL(id=21)
	TS	EXTERNAL(id=22)
	NP	EXTERNAL(id=23)
	NF	EXTERNAL(id=24)
	RL	EXTERNAL(id=25)
	FN	EXTERNAL(id=26)
	FC	EXTERNAL(id=27)
	FP	EXTERNAL(id=28)
	DL	EXTERNAL(id=29)
	BA	EXTERNAL(id=30)
	BS	EXTERNAL(id=31)
	TL	EXTERNAL(id=32)
	RI	HUFFMAN(codes={0},lengths={0})
	HC	EXTERNAL(id=36)
	BB	BYTE_ARRAY_LEN(len_codec={EXTERNAL(id=42)},val_codec={EXTERNAL(id=37)}
	XAZ	BYTE_ARRAY_STOP(stop=9,id=5783898)
	ASC	BYTE_ARRAY_LEN(len_codec={HUFFMAN(codes={1},lengths={0})},val_codec={EXTERNAL(id=4281155)}
	SAZ	BYTE_ARRAY_STOP(stop=9,id=5456218)
	XSC	BYTE_ARRAY_LEN(len_codec={HUFFMAN(codes={1},lengths={0})},val_codec={EXTERNAL(id=5788483)}
```
```

### CODE REVIEW

**Comment by jkbonfield** [comment]
I'm now happy for this to be reviewed, but it'll fail tests until htslib bit gets merged.

File: cram_size.c [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+
+    kh_destroy(cu, cu_size);
+    cram_cid2ds_free(cid2ds);
+
+    // This is a little bit naughty as it assumes that the hFILE pointer
+    // is the very first thing in the opaque cram_fd struct, which it is.
+    // The alternative would be to add an explicit htslib CRAM API member
+    // function to return the internal hFILE handle, but if we expose this
+    // then people may feel it's fair game to start modifying it.
+    off_t end = htell(*(hFILE **)in->fp.cram);
```
Comment (daviesrob) [inline cram_size.c:555]: This naughtiness could be avoided by opening the input using `hopen()` and then using `hts_hopen()` to upgrade it to a `samFile`.  That way you would still have a pointer to the underlying `hFILE` to pass to `htell()`.

Comment (jkbonfield) [reply cram_size.c:555]: Oh that's useful!  I was wondering how to get an hfile out and was rather surprised it wasn't possible, but this trick avoids the necessity.

File: bamtk.c [hunk 2]
```diff
@@ -289,6 +291,7 @@ int main(int argc, char *argv[])
     else if (strcmp(argv[1], "samples") == 0)     ret = main_samples(argc-1, argv+1);
     else if (strcmp(argv[1], "consensus") == 0) ret = main_consensus(argc-1, argv+1);
     else if (strcmp(argv[1], "reference") == 0) ret = main_reference(argc-1, argv+1);
+    else if (strcmp(argv[1], "cram_size") == 0) ret = main_cram_size(argc-1, argv+1);
```
[the remark this answers is not in the document]
Comment (jkbonfield) [reply bamtk.c:294]: I see bcftools plugins have many foo-bar syntax, so I'll just rename it.

**Comment by daviesrob** [comment]
Oops, I spotted a couple of minor problems after merging this.  See #1785

**Pull request merged**